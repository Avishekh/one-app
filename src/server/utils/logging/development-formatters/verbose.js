/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import util from 'util';

import dropEntryBasedOnLevel from '../level-dropper';
import {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
} from './utils';

const formatter = (levelName, ...args) => {
  if (dropEntryBasedOnLevel(levelName)) {
    return null;
  }

  const level = coloredLevels[levelName];
  const obj = args[0];

  switch (obj && obj.type) {
    case 'request':
      if (obj.request.direction === 'out') {
        return `${level}: 💻 ➡ 🗄️ ${printStatusCode(obj)} ${printStatusMessage(obj)} ${obj.request.metaData.method} ${obj.request.address.uri} ${printDurationTime(obj)}ms`;
      }
      return `${level}: 🌍 ➡ 💻 ${printStatusCode(obj)} ${printStatusMessage(obj)} ${obj.request.metaData.method} ${obj.request.address.uri} ${obj.request.timings.ttfb}/${obj.request.timings.duration}ms`;
    default:
      return `${level}: ${util.format.apply(null, args)}`;
  }
};

const beforeWrite = undefined;
const afterWrite = undefined;

export {
  // eslint rule prefer-default doesn't work for this pattern, so we'll be explicit
  beforeWrite,
  afterWrite,
  formatter,
};
