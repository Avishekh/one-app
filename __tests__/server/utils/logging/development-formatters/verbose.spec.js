/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { beforeWrite, afterWrite, formatter } from '../../../../../src/server/utils/logging/development-formatters/verbose';

jest.mock('chalk');

describe('verbose', () => {
  describe('beforeWrite', () => {
    it('is undefined', () => {
      expect(beforeWrite).toBe(undefined);
    });
  });

  describe('afterWrite', () => {
    it('is undefined', () => {
      expect(afterWrite).toBe(undefined);
    });
  });

  describe('formatter', () => {
    it('prints the level with color', () => {
      expect(formatter('error', 'hello')).toMatchSnapshot();
    });

    it('drops lower levels', () => {
      expect(formatter('debug', 1, 2, 3)).toBe(null);
    });

    it('pretty-prints an incoming request', () => {
      const entry = {
        type: 'request',
        request: {
          direction: 'in',
          protocol: 'https',
          address: {
            uri: 'https://example.com/resource',
          },
          metaData: {
            method: 'GET',
            correlationId: '123',
            // null to explicitly signal no value, undefined if not expected for every request
            host: 'example.com',
            referrer: 'https://example.com/other-resource',
            userAgent: 'Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)',
            location: undefined,
          },
          timings: {
            duration: 13,
            ttfb: 12,
          },
          statusCode: 200,
          statusText: 'OK',
        },
      };
      expect(formatter('error', entry)).toMatchSnapshot();
    });

    it('pretty-prints an outgoing request', () => {
      const entry = {
        type: 'request',
        request: {
          direction: 'out',
          protocol: 'https',
          address: {
            uri: 'https://example.com/resource',
          },
          metaData: {
            method: 'GET',
            // null to explicitly signal no value, undefined if not expected for every request
            correlationId: '123',
          },
          timings: {
            duration: 13,
          },
          statusCode: 200,
          statusText: 'OK',
        },
      };
      expect(formatter('error', entry)).toMatchSnapshot();
    });

    describe('unknown type', () => {
      it('is formatted by util.format', () => {
        expect(formatter('error', 'hello, %s', 'world')).toMatchSnapshot();
      });
    });
  });
});
