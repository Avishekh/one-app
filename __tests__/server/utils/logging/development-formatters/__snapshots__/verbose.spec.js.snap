// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verbose formatter pretty-prints an incoming request 1`] = `"<red>error</red>: 🌍 ➡ 💻 <green>200</green> OK GET https://example.com/resource 12/13ms"`;

exports[`verbose formatter pretty-prints an outgoing request 1`] = `"<red>error</red>: 💻 ➡ 🗄️ <green>200</green> OK GET https://example.com/resource <green>13</green>ms"`;

exports[`verbose formatter prints the level with color 1`] = `"<red>error</red>: hello"`;

exports[`verbose formatter unknown type is formatted by util.format 1`] = `"<red>error</red>: hello, world"`;
